src/ziggurat/kafka_consumer/consumer_handler.clj

```diff
@@ -0,0 +1,85 @@
+   (when (> (count batch) 0)
+     (let [message (map->BatchPayload {:batch         batch
+                                       :retry-count        current-retry-count
+                                       :topic-entity topic-entity})]
+       (producer/retry message)))))
+
+(defn process
+  [batch-handler batch-payload]
+  (let [batch               (:batch batch-payload)
+        topic-entity        (:topic-entity batch-payload)
+        current-retry-count (:retry-count batch-payload)
+        batch-size          (count batch)]
+    (try
+      (when (not-empty batch)
+        (log/infof "[Consumer Group: %s] Processing the batch with %d messages" topic-entity batch-size)
+        (let [start-time             (Instant/now)
+              result                 (batch-handler batch)
+              time-taken-in-millis   (.toMillis (Duration/between start-time (Instant/now)))]
+          (when (map? result)
+            (let [messages-to-be-retried (:retry result)
+                  to-be-retried-count    (count messages-to-be-retried)
+                  skip-count             (count (:skip result))
+                  success-count          (- batch-size (+ to-be-retried-count skip-count))]
+              (log/infof "[Consumer Group: %s] Processed the batch with success: [%d], skip: [%d] and retries: [%d] \n"
+                         topic-entity success-count skip-count to-be-retried-count)
+              (publish-batch-process-metrics topic-entity batch-size success-count skip-count to-be-retried-count time-taken-in-millis)
+              (retry messages-to-be-retried current-retry-count topic-entity)))))
+      (catch Exception e
+        (do
+          (metrics/increment-count batch-consumption-metric-ns "exception" 1 {:topic-entity topic-entity})
+          (log/errorf e "[Consumer Group: %s] Exception received while processing messages \n" topic-entity)
+          (retry batch-payload))))))
+
+(defn- commit-offsets
+  [consumer topic-entity]
+  (try
+    (.commitSync consumer)
+    (catch Exception e
+      (metrics/increment-count batch-consumption-metric-ns "commit.failed.exception" 1 {:topic-entity topic-entity})
+      (log/error "Exception while committing offsets:" e))))
+
+(defn- create-batch-payload
+  [records topic-entity]
+  (let [key-value-pairs (map (fn [^ConsumerRecord m]
+                               {:value (.value m) :key (.key m)}) records)]
+    (map->BatchPayload {:batch key-value-pairs :topic-entity topic-entity})))
+
+(defn poll-for-messages
+  [^Consumer consumer handler-fn topic-entity consumer-config]
+  (try
+    (loop [records []]
+      (let [batch-payload (create-batch-payload records topic-entity)]
+        (process handler-fn batch-payload))
+      (commit-offsets consumer topic-entity)
+      (recur (seq (.poll consumer (Duration/ofMillis (:poll-timeout consumer-config))))))
+    (catch WakeupException e)
+    (finally (do (log/info "Closing the Kafka Consumer for: " topic-entity)
```

Comment:
Why are we not closing the consumer within the catch block itself?

Reply:
In Java closing any resource should be done in finally, because finally block will run all time.

Reply:
So, for any exception other than WakeupException, we will be silently closing the consumer?

Reply:
For all exceptions, the code will enter the finally block. But, for WakeupException, it will remain silent. While other exceptions will get thrown (and not be caught).

We're catching WakeupException explicitly because we know it will be throw when consumers are stopped by making a .wakeup call.